File: src/modules/pronunciation/ctc.ts
import { stripStress } from "./g2p";
import type { AcousticFrame, AcousticLabels, PronunciationPhonemeScore } from "./types";

const BLANK_LABEL = "<blank>";
const IMPOSSIBLE = Number.NEGATIVE_INFINITY;
const LOG_PROB_FALLBACK = -14;
const BLANK_PENALTY = 6;

export interface AlignmentResult {
  phonemes: PronunciationPhonemeScore[];
  averageLogProb: number;
}

/**
 * CTC Viterbi alignment over a per-frame log-probability matrix.
 *
 * The DP table is flat (`Float32Array(rows * cols)`) and backtracking
 * indices live in an `Int32Array` — no nested-array allocs, no string
 * lookups in the inner loop. Both target phonemes and frame log-probs
 * are indexed by integers into `labels.labels`, which is the same ABI
 * the WASM port (Phase G) uses.
 */
export function ctcViterbiAlign({
  target,
  frames,
  labels,
}: {
  target: string[];
  frames: AcousticFrame[];
  labels: AcousticLabels;
}): AlignmentResult {
  if (target.length === 0 || frames.length === 0) {
    return { phonemes: [], averageLogProb: IMPOSSIBLE };
  }

  const phonemeIndex = buildPhonemeIndex(labels);
  const targetIndices = target.map((phoneme) => phonemeIndex.get(stripStress(phoneme)) ?? -1);
  const labelIndices = buildCtcLabelIndices(targetIndices, labels.blankIndex);
  const cols = labelIndices.length;
  const rows = frames.length;
  const firstFrame = frames[0];
  if (!firstFrame) return { phonemes: [], averageLogProb: IMPOSSIBLE };

  const dp = new Float32Array(rows * cols);
  const back = new Int32Array(rows * cols);
  dp.fill(IMPOSSIBLE);

  dp[0] = readLogProb(firstFrame, labelIndices[0] ?? labels.blankIndex, labels.blankIndex);
  if (cols > 1) {
    dp[1] = readLogProb(firstFrame, labelIndices[1] ?? labels.blankIndex, labels.blankIndex);
  }

  for (let t = 1; t < rows; t += 1) {
    const frame = frames[t] ?? firstFrame;
    const rowOffset = t * cols;
    const prevOffset = (t - 1) * cols;
    for (let s = 0; s < cols; s += 1) {
      let bestPrev = s;
      let bestScore = dp[prevOffset + s] ?? IMPOSSIBLE;
      const prev1 = s > 0 ? (dp[prevOffset + s - 1] ?? IMPOSSIBLE) : IMPOSSIBLE;
      if (prev1 > bestScore) {
        bestScore = prev1;
        bestPrev = s - 1;
      }
      const skipAllowed =
        s > 1 && labelIndices[s] !== labels.blankIndex && labelIndices[s] !== labelIndices[s - 2];
      const prev2 = skipAllowed ? (dp[prevOffset + s - 2] ?? IMPOSSIBLE) : IMPOSSIBLE;
      if (prev2 > bestScore) {
        bestScore = prev2;
        bestPrev = s - 2;
      }
      const labelIdx = labelIndices[s] ?? labels.blankIndex;
      dp[rowOffset + s] = bestScore + readLogProb(frame, labelIdx, labels.blankIndex);
      back[rowOffset + s] = bestPrev;
    }
  }

  let finalState = cols - 1;
  const finalAlt = cols - 2;
  if (
    finalAlt >= 0 &&
    (dp[(rows - 1) * cols + finalAlt] ?? IMPOSSIBLE) >
      (dp[(rows - 1) * cols + finalState] ?? IMPOSSIBLE)
  ) {
    finalState = finalAlt;
  }

  const statePath = new Int32Array(rows);
  let state = finalState;
  for (let t = rows - 1; t >= 0; t -= 1) {
    statePath[t] = state;
    state = back[t * cols + state] ?? state;
  }

  const phonemes = targetIndices.map((labelIdx, index) => {
    const ctcStateIdx = index * 2 + 1;
    let first = -1;
    let last = -1;
    for (let t = 0; t < rows; t += 1) {
      if (statePath[t] === ctcStateIdx) {
        if (first === -1) first = t;
        last = t;
      }
    }
    if (first === -1) {
      const fallback = Math.min(rows - 1, Math.round((index / target.length) * rows));
      first = fallback;
      last = fallback;
    }

    let sum = 0;
    for (let t = first; t <= last; t += 1) {
      const frame = frames[t];
      if (!frame) continue;
      sum += readLogProb(frame, labelIdx, labels.blankIndex);
    }
    const span = Math.max(1, last - first + 1);
    const avg = sum / span;
    const phoneme = stripStress(target[index] ?? "");
    const detected = mostLikelyLabel(frames, first, last, labels);
    const score = logProbToScore(avg);
    return {
      phoneme,
      expectedIndex: index,
      startMs: frames[first]?.timeMs ?? 0,
      endMs: (frames[last]?.timeMs ?? frames[first]?.timeMs ?? 0) + 20,
      score,
      detectedPhoneme: detected === BLANK_LABEL ? null : detected,
      issue: score >= 76 ? "ok" : detected && detected !== phoneme ? "substitution" : "weak",
    } satisfies PronunciationPhonemeScore;
  });

  const lastRow = (rows - 1) * cols;
  const best = Math.max(
    dp[lastRow + cols - 1] ?? IMPOSSIBLE,
    cols >= 2 ? (dp[lastRow + cols - 2] ?? IMPOSSIBLE) : IMPOSSIBLE,
  );
  return { phonemes, averageLogProb: best / rows };
}

function buildPhonemeIndex(labels: AcousticLabels): Map<string, number> {
  const out = new Map<string, number>();
  for (let i = 0; i < labels.labels.length; i += 1) {
    const label = labels.labels[i];
    if (label) out.set(label, i);
  }
  return out;
}

function buildCtcLabelIndices(targetIndices: number[], blankIndex: number): Int32Array {
  const out = new Int32Array(targetIndices.length * 2 + 1);
  out[0] = blankIndex;
  for (let i = 0; i < targetIndices.length; i += 1) {
    const target = targetIndices[i] ?? blankIndex;
    out[i * 2 + 1] = target >= 0 ? target : blankIndex;
    out[i * 2 + 2] = blankIndex;
  }
  return out;
}

function readLogProb(frame: AcousticFrame, labelIdx: number, blankIndex: number): number {
  if (labelIdx < 0 || labelIdx >= frame.logProbs.length) {
    const blank = frame.logProbs[blankIndex];
    return typeof blank === "number" ? blank - BLANK_PENALTY : LOG_PROB_FALLBACK;
  }
  const direct = frame.logProbs[labelIdx];
  if (typeof direct === "number" && Number.isFinite(direct)) return direct;
  const blank = frame.logProbs[blankIndex];
  return typeof blank === "number" ? blank - BLANK_PENALTY : LOG_PROB_FALLBACK;
}

function mostLikelyLabel(
  frames: AcousticFrame[],
  first: number,
  last: number,
  labels: AcousticLabels,
): string | null {
  const labelCount = labels.labels.length;
  if (labelCount === 0) return null;
  const totals = new Float64Array(labelCount);
  for (let t = first; t <= last; t += 1) {
    const frame = frames[t];
    if (!frame) continue;
    for (let i = 0; i < labelCount; i += 1) {
      const value = frame.logProbs[i] ?? 0;
      totals[i] = (totals[i] ?? 0) + value;
    }
  }
  let bestIdx = -1;
  let bestScore = IMPOSSIBLE;
  for (let i = 0; i < labelCount; i += 1) {
    if ((totals[i] ?? IMPOSSIBLE) > bestScore) {
      bestIdx = i;
      bestScore = totals[i] ?? IMPOSSIBLE;
    }
  }
  return bestIdx >= 0 ? (labels.labels[bestIdx] ?? null) : null;
}

function logProbToScore(logProb: number): number {
  if (!Number.isFinite(logProb)) return 0;
  return Math.max(0, Math.min(100, Math.round(100 + logProb * 18)));
}

File: core/vocab-core/src/ctc.rs
pub const IMPOSSIBLE: f32 = f32::NEG_INFINITY;

const LOG_PROB_FALLBACK: f32 = -14.0;
const BLANK_PENALTY: f32 = 6.0;

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct AlignmentSpan {
    pub start_frame: usize,
    pub end_frame: usize,
    pub avg_log_prob: f32,
    pub detected_index: i32,
}

#[derive(Clone, Debug, PartialEq)]
pub struct AlignmentResult {
    pub spans: Vec<AlignmentSpan>,
    pub average_log_prob: f32,
}

impl AlignmentResult {
    pub fn empty() -> Self {
        Self {
            spans: Vec::new(),
            average_log_prob: IMPOSSIBLE,
        }
    }
}

pub fn align_indices(
    target: &[i32],
    blank_index: i32,
    frames: &[f32],
    frame_count: usize,
    label_count: usize,
) -> AlignmentResult {
    if target.is_empty() || frame_count == 0 || label_count == 0 {
        return AlignmentResult::empty();
    }
    let Some(expected_frame_values) = frame_count.checked_mul(label_count) else {
        return AlignmentResult::empty();
    };
    if frames.len() < expected_frame_values {
        return AlignmentResult::empty();
    }

    let blank = if blank_index >= 0 && (blank_index as usize) < label_count {
        blank_index
    } else {
        0
    };
    let target_len = target.len();
    let cols = target_len * 2 + 1;

    let mut labels: Vec<i32> = Vec::with_capacity(cols);
    labels.push(blank);
    for &phoneme in target {
        labels.push(if phoneme >= 0 { phoneme } else { blank });
        labels.push(blank);
    }

    let mut dp = vec![IMPOSSIBLE; frame_count * cols];
    let mut back = vec![0i32; frame_count * cols];

    dp[0] = log_prob(frames, label_count, 0, labels[0], blank);
    if cols > 1 {
        dp[1] = log_prob(frames, label_count, 0, labels[1], blank);
    }

    for t in 1..frame_count {
        let row_offset = t * cols;
        let prev_offset = (t - 1) * cols;
        for s in 0..cols {
            let mut best_prev = s as i32;
            let mut best_score = dp[prev_offset + s];
            if s > 0 {
                let cand = dp[prev_offset + s - 1];
                if cand > best_score {
                    best_score = cand;
                    best_prev = (s - 1) as i32;
                }
            }
            let skip_allowed = s > 1 && labels[s] != blank && labels[s] != labels[s - 2];
            if skip_allowed {
                let cand = dp[prev_offset + s - 2];
                if cand > best_score {
                    best_score = cand;
                    best_prev = (s - 2) as i32;
                }
            }
            dp[row_offset + s] = best_score + log_prob(frames, label_count, t, labels[s], blank);
            back[row_offset + s] = best_prev;
        }
    }

    let last_row = (frame_count - 1) * cols;
    let mut final_state = cols - 1;
    if cols >= 2 {
        let alt = cols - 2;
        if dp[last_row + alt] > dp[last_row + final_state] {
            final_state = alt;
        }
    }

    let mut state_path: Vec<i32> = vec![0; frame_count];
    let mut state = final_state;
    for t in (0..frame_count).rev() {
        state_path[t] = state as i32;
        state = back[t * cols + state] as usize;
    }

    let mut totals = vec![0.0f64; label_count];
    let spans = target
        .iter()
        .enumerate()
        .map(|(i, &phoneme)| {
            let ctc_state = (i * 2 + 1) as i32;
            let mut first: i64 = -1;
            let mut last: i64 = -1;
            for (t, state) in state_path.iter().enumerate() {
                if *state == ctc_state {
                    if first < 0 {
                        first = t as i64;
                    }
                    last = t as i64;
                }
            }
            if first < 0 {
                let fallback =
                    (((i as f32) / (target_len as f32)) * (frame_count as f32)).round() as i64;
                let clamped = fallback.min((frame_count as i64) - 1).max(0);
                first = clamped;
                last = clamped;
            }

            let phoneme_for_span = if phoneme >= 0 { phoneme } else { blank };
            let mut sum = 0.0f32;
            totals.fill(0.0);

            for t in (first as usize)..=(last as usize) {
                sum += log_prob(frames, label_count, t, phoneme_for_span, blank);
                let offset = t * label_count;
                for j in 0..label_count {
                    totals[j] += frames[offset + j] as f64;
                }
            }

            let span = ((last - first + 1).max(1)) as f32;
            let avg = sum / span;

            let mut detected_index: i32 = -1;
            let mut best_total = f64::NEG_INFINITY;
            for (j, total) in totals.iter().enumerate() {
                if *total > best_total {
                    best_total = *total;
                    detected_index = j as i32;
                }
            }

            AlignmentSpan {
                start_frame: first as usize,
                end_frame: last as usize,
                avg_log_prob: avg,
                detected_index,
            }
        })
        .collect();

    let best = (dp[last_row + cols - 1]).max(if cols >= 2 {
        dp[last_row + cols - 2]
    } else {
        IMPOSSIBLE
    });

    AlignmentResult {
        spans,
        average_log_prob: best / frame_count as f32,
    }
}

#[inline]
fn log_prob(frames: &[f32], label_count: usize, frame_idx: usize, label: i32, blank: i32) -> f32 {
    if label < 0 || (label as usize) >= label_count {
        let blank_val = frames
            .get(frame_idx * label_count + blank as usize)
            .copied()
            .unwrap_or(LOG_PROB_FALLBACK);
        return if blank_val.is_finite() {
            blank_val - BLANK_PENALTY
        } else {
            LOG_PROB_FALLBACK
        };
    }
    let v = frames[frame_idx * label_count + label as usize];
    if v.is_finite() {
        return v;
    }
    let blank_val = frames
        .get(frame_idx * label_count + blank as usize)
        .copied()
        .unwrap_or(LOG_PROB_FALLBACK);
    if blank_val.is_finite() {
        blank_val - BLANK_PENALTY
    } else {
        LOG_PROB_FALLBACK
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deterministic_short_word_matches_expected_spans() {
        let target = [1, 2, 3];
        let label_count = 4;
        let frames = deterministic_frames(&target, label_count);
        let result = align_indices(&target, 0, &frames, frames.len() / label_count, label_count);

        assert_eq!(result.spans.len(), target.len());
        assert_span(result.spans[0], 0, 3, 1);
        assert_span(result.spans[1], 5, 8, 2);
        assert_span(result.spans[2], 10, 13, 3);
        assert!(result.average_log_prob.is_finite());
    }

    #[test]
    fn empty_inputs_return_empty_alignment() {
        let result = align_indices(&[], 0, &[], 0, 2);

        assert!(result.spans.is_empty());
        assert_eq!(result.average_log_prob, IMPOSSIBLE);
    }

    #[test]
    fn invalid_frame_shape_returns_empty_alignment() {
        let result = align_indices(&[1], 0, &[-0.1, -0.2], 2, 2);

        assert!(result.spans.is_empty());
    }

    fn deterministic_frames(target: &[i32], label_count: usize) -> Vec<f32> {
        let mut frames = Vec::new();
        for &phoneme in target {
            for i in 0..4 {
                let mut frame = vec![-9.0; label_count];
                frame[0] = -5.0;
                frame[phoneme as usize] = if i == 0 || i == 3 { -1.4 } else { -0.22 };
                frames.extend(frame);
            }
            let mut trail = vec![-9.0; label_count];
            trail[0] = -0.12;
            trail[phoneme as usize] = -2.5;
            frames.extend(trail);
        }
        frames
    }

    fn assert_span(span: AlignmentSpan, start: usize, end: usize, detected: i32) {
        assert_eq!(span.start_frame, start);
        assert_eq!(span.end_frame, end);
        assert_eq!(span.detected_index, detected);
        assert!(span.avg_log_prob.is_finite());
    }
}
